sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala

```diff
@@ -42,7 +45,15 @@ class Iso8601DateFormatter(
   override def parse(s: String): Int = {
     val specialDate = convertSpecialDate(s.trim, zoneId)
     specialDate.getOrElse {
-      val localDate = LocalDate.parse(s, formatter)
+      val localDate = try {
+        LocalDate.parse(s, formatter)
+      } catch {
+        case e: DateTimeParseException if DateFormatter.hasDiffResult(s, pattern, zoneId) =>
+          throw new RuntimeException(e.getMessage + ", set " +
```

Comment:
IIRC some places just call formatter and catch `NonFatal`. Can you do some checks and make sure we don't catch non-fatal blindly?

Reply:
Thanks for reminding, fix for the non-codegen in fbac26d.